source/tests/system/nirfmxspecan_driver_api_tests.cpp

```diff
@@ -106,10 +160,153 @@ TEST_F(NiRFmxSpecAnDriverApiTests, SpectrumBasicFromExample_DataLooksReasonable)
+  EXPECT_SUCCESS(session, am_to_am_response);
+}
+
+// Note: there aren't any complex attributes in attributes.py, but this at least exercises the code.
+TEST_F(NiRFmxSpecAnDriverApiTests, SetAttributeComplex_ExpectedError)
+{
+  const auto session = init_session(stub(), PXI_5663);
+
+  EXPECT_RFMX_ERROR(
+      -380231, "This attribute is read-only and cannot be written", session,
+      client::set_attribute_ni_complex_double_array(stub(), session, "", NiRFmxSpecAnAttribute::NIRFMXSPECAN_ATTRIBUTE_SEM_RESULTS_CARRIER_FREQUENCY, complex_number_array({1.2, 2.2}, {1e6, 1.01e6})));
+}
+
+// Note: there aren't any i8 attributes in attributes.py, but this at least exercises the code.
+TEST_F(NiRFmxSpecAnDriverApiTests, SetAttributeInt8_ExpectedError)
+{
+  const auto session = init_session(stub(), PXI_5663);
+
+  EXPECT_RFMX_ERROR(
+      -380251, "Incorrect data type specified", session,
+      client::set_attribute_i8(stub(), session, "", NiRFmxSpecAnAttribute::NIRFMXSPECAN_ATTRIBUTE_NF_EXTERNAL_PREAMP_PRESENT, 1));
+  EXPECT_RFMX_ERROR(
+      -380251, "Incorrect data type specified", session,
+      client::set_attribute_u8(stub(), session, "", NiRFmxSpecAnAttribute::NIRFMXSPECAN_ATTRIBUTE_NF_EXTERNAL_PREAMP_PRESENT, 1));
+  EXPECT_RFMX_ERROR(
+      -380251, "Incorrect data type specified", session,
+      client::set_attribute_i8_array(stub(), session, "", NiRFmxSpecAnAttribute::NIRFMXSPECAN_ATTRIBUTE_NF_EXTERNAL_PREAMP_PRESENT, {1, 0, -1, 0}));
+  EXPECT_RFMX_ERROR(
+      -380251, "Incorrect data type specified", session,
+      client::set_attribute_u8_array(stub(), session, "", NiRFmxSpecAnAttribute::NIRFMXSPECAN_ATTRIBUTE_NF_EXTERNAL_PREAMP_PRESENT, {1, 0, 1, 0}));
+}
+
+// Note: there aren't any i16 attributes in attributes.py, but this at least exercises the code.
+TEST_F(NiRFmxSpecAnDriverApiTests, SetAttributeInt16_ExpectedError)
+{
+  const auto session = init_session(stub(), PXI_5663);
+
+  EXPECT_RFMX_ERROR(
+      -380251, "Incorrect data type specified", session,
+      client::set_attribute_i16(stub(), session, "", NiRFmxSpecAnAttribute::NIRFMXSPECAN_ATTRIBUTE_NF_EXTERNAL_PREAMP_PRESENT, -400));
+  EXPECT_RFMX_ERROR(
+      -380251, "Incorrect data type specified", session,
+      client::set_attribute_u16(stub(), session, "", NiRFmxSpecAnAttribute::NIRFMXSPECAN_ATTRIBUTE_NF_EXTERNAL_PREAMP_PRESENT, 400));
+}
+
 TEST_F(NiRFmxSpecAnDriverApiTests, SetAndGetAttributeInt32_Succeeds)
 {
   auto session = init_session(stub(), PXI_5663);
-  EXPECT_SUCCESS(session, client::set_attribute_i32(stub(), session, "", NiRFmxSpecAnAttribute::NIRFMXSPECAN_ATTRIBUTE_NF_EXTERNAL_PREAMP_PRESENT, NiRFmxSpecAnInt32AttributeValues::NIRFMXSPECAN_INT32_NF_EXTERNAL_PREAMP_PRESENT_TRUE));
+  EXPECT_SUCCESS(
+      session,
+      client::set_attribute_i32(stub(), session, "", NiRFmxSpecAnAttribute::NIRFMXSPECAN_ATTRIBUTE_NF_EXTERNAL_PREAMP_PRESENT, NiRFmxSpecAnInt32AttributeValues::NIRFMXSPECAN_INT32_NF_EXTERNAL_PREAMP_PRESENT_TRUE));
```

Comment:
Hmm, I thought we got rid of all the "TRUE" enum values?

Reply:
Attributes are different. That's a big "attribute_values" enum that includes bool enums as well as non-bool enums. So the simple swap doesn't work.

We could add a `get/set_attribute_bool` wrapper that calls into `get/set_attribute_i32`. I think that would be cool, although nothing would stop you from calling it for some other int32 attribute that's not a bool (though is that worse than allowing passing in cross-enums?).

Reply:
Yeah, the bool wrapper does sound kinda nice but I'm not sure it's worth the complication...